File: edge/api-gateway/internal/problem/problem.go
// Package problem writes the RFC 9457 documents the gateway refuses a request with.
//
// The shape matches the one services/ledger-api emits, field for field and namespace for namespace,
// because a client sees both and should not need to know which tier answered. The type URI is the
// part a client is allowed to branch on; title and detail may be reworded freely.
//
// A detail string is written for a human and read by an attacker. Nothing here explains which of a
// token's claims failed, whether an account exists, or what the ledger said - the log line carries
// that, and the log line is not served to the caller.
package problem

import (
	"encoding/json"
	"net/http"

	"github.com/k-napiontek/tessera-bank/edge/api-gateway/internal/correlation"
)

// The namespace is the estate's, not the gateway's. One vocabulary of problem types spans the
// tiers, and a client matching on a prefix matches both.
const namespace = "https://problems.tesserabank.example/"

// Type is a problem kind: the stable URI, and the title that goes with it.
type Type struct {
	slug  string
	title string
}

// URI is the machine-readable identifier a client branches on.
func (t Type) URI() string { return namespace + t.slug }

// Title is the human-readable summary.
func (t Type) Title() string { return t.title }

// The kinds the edge can produce. Business failures belong to the ledger and are relayed from it
// untouched; nothing here duplicates one.
var (
	Unauthenticated   = Type{"unauthenticated", "Authentication required"}
	Forbidden         = Type{"forbidden", "Not permitted"}
	RateLimited       = Type{"rate-limited", "Too many requests"}
	PayloadTooLarge   = Type{"payload-too-large", "Request body is too large"}
	NoRoute           = Type{"no-route", "No such route"}
	UpstreamTimeout   = Type{"upstream-timeout", "The ledger did not answer in time"}
	UpstreamUnusable  = Type{"upstream-unusable", "The ledger could not be reached"}
	UpstreamOversized = Type{"upstream-oversized", "The ledger's response is too large to relay"}
)

// document is the wire form, matching the Problem schema in contracts/openapi/ledger-core.yaml.
type document struct {
	Type          string `json:"type"`
	Title         string `json:"title"`
	Status        int    `json:"status"`
	Detail        string `json:"detail,omitempty"`
	Instance      string `json:"instance,omitempty"`
	CorrelationID string `json:"correlationId,omitempty"`
}

// Write replaces whatever was going to be sent with a problem document.
func Write(w http.ResponseWriter, r *http.Request, status int, kind Type, detail string) {
	body, err := json.Marshal(document{
		Type:     kind.URI(),
		Title:    kind.Title(),
		Status:   status,
		Detail:   detail,
		Instance: r.URL.Path,
		// The resolved id, so a request that arrived without one is still traceable and a caller
		// who sent something that was not a UUID does not get it echoed back at them.
		CorrelationID: correlation.FromContext(r.Context()),
	})
	if err != nil {
		// A struct of strings and an int cannot fail to marshal; if it somehow does, an empty 500
		// is still better than a half-written body.
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "application/problem+json")
	w.Header().Set("Cache-Control", "no-store")
	w.WriteHeader(status)
	_, _ = w.Write(body)
}

File: edge/web-banking/src/api/problem.ts
/**
 * RFC 9457 Problem Details, read from two producers that do not share a contract.
 *
 * `contracts/openapi/ledger-core.yaml` declares the shape and the ledger's nine problem types. The
 * gateway emits eight more of its own - `unauthenticated`, `forbidden`, `rate-limited`,
 * `payload-too-large`, `no-route`, `upstream-timeout`, `upstream-unusable`, `upstream-oversized` -
 * which follow-up F-34 records as being in no contract at all. This module is the second consumer
 * to need them, which is the argument for writing that contract rather than a reason to guess.
 *
 * `type` is the stable identifier; `title` and `detail` are prose the producers may reword without
 * a contract change. So the wording a customer sees is keyed on `type` and falls back to `title` -
 * never to a bare status code, which tells a customer nothing and a support engineer little more.
 */

export const PROBLEM_NAMESPACE = 'https://problems.tesserabank.example/';

export interface Violation {
  readonly field: string;
  readonly message: string;
}

export interface Problem {
  /** Full `type` URI as it arrived, or '' when the body was not a problem document. */
  readonly type: string;
  /** The last segment of `type` - what the code branches on. '' when there is none. */
  readonly slug: string;
  readonly title: string;
  readonly status: number;
  readonly detail: string | undefined;
  readonly correlationId: string | undefined;
  readonly violations: readonly Violation[];
}

function asRecord(value: unknown): Record<string, unknown> {
  return typeof value === 'object' && value !== null ? (value as Record<string, unknown>) : {};
}

function asString(value: unknown): string | undefined {
  return typeof value === 'string' ? value : undefined;
}

function readViolations(value: unknown): readonly Violation[] {
  if (!Array.isArray(value)) {
    return [];
  }
  const violations: Violation[] = [];
  for (const entry of value) {
    const record = asRecord(entry);
    const field = asString(record['field']);
    const message = asString(record['message']);
    if (field !== undefined && message !== undefined) {
      violations.push({ field, message });
    }
  }
  return violations;
}

/**
 * Read a problem document out of a response body.
 *
 * The body is whatever arrived. A gateway timeout served by infrastructure the estate does not own
 * is HTML, and a screen still has to say something honest about it, so nothing here assumes the
 * shape it hopes for.
 */
export function parseProblem(body: unknown, status: number): Problem {
  const record = asRecord(body);
  const type = asString(record['type']) ?? '';
  const slug = type.startsWith(PROBLEM_NAMESPACE) ? type.slice(PROBLEM_NAMESPACE.length) : '';
  const declaredStatus = record['status'];
  return {
    type,
    slug,
    title: asString(record['title']) ?? '',
    status: typeof declaredStatus === 'number' ? declaredStatus : status,
    detail: asString(record['detail']),
    correlationId: asString(record['correlationId']),
    violations: readViolations(record['violations']),
  };
}

/**
 * What the customer is told.
 *
 * Deliberately plain, and deliberately without the type URI, the status code or the correlation id
 * - the first two mean nothing to the person reading, and the third belongs beside a "contact us"
 * affordance rather than inside the sentence explaining what happened.
 */
const WORDING: Readonly<Record<string, string>> = Object.freeze({
  // The ledger's own.
  'insufficient-funds': 'There is not enough available balance on that account to cover this.',
  'currency-mismatch': 'Both accounts must be in the same currency. This estate performs no conversion.',
  'not-found': 'We could not find that account or transfer.',
  'validation-failed': 'Some of the details are not valid. Check the highlighted fields.',
  'idempotency-conflict': 'This looks like a different transfer sent under the same reference. Start it again.',
  'conflicting-state': 'That has already been done, or it can no longer be done.',
  'account-already-open': 'That account reference is already in use.',
  'not-actionable': 'We understood the request but cannot carry it out as it stands.',
  // Never "nothing was changed" - that is a claim about the ledger this application cannot make.
  internal: 'Something went wrong at our end, and we do not yet know whether this went through.',
  // The gateway's own - F-34.
  unauthenticated: 'Your session is not valid any more. Sign in again.',
  forbidden: 'This account is not one you are allowed to see.',
  'rate-limited': 'Too many requests just now. Wait a moment and try again.',
  'payload-too-large': 'That request was too large to send.',
  'no-route': 'That is not something this service offers.',
  'upstream-timeout': 'The bank did not answer in time. Your request may or may not have gone through.',
  'upstream-unusable': 'We could not reach the bank. Your request may or may not have gone through.',
  'upstream-oversized': 'The answer was too large to show.',
});

export function describeProblem(problem: Problem): string {
  const known = WORDING[problem.slug];
  if (known !== undefined) {
    return known;
  }
  if (problem.title !== '') {
    return problem.title;
  }
  return 'Something went wrong, and we could not tell what.';
}

/**
 * True when the outcome of a money-moving request is genuinely unknown to the client.
 *
 * **The line is 4xx against 5xx, not a list of problem types.** A 4xx is decided before any money
 * moves - validation, funds, a conflicting state, an expired token - so it is the only class of
 * answer a client may present to a customer as "not sent". A 5xx is not an answer about the
 * ledger at all: the request may have committed and lost its response on the way back, and
 * anything standing between the customer and the ledger - a reverse proxy, a load balancer, the
 * gateway itself - can produce one carrying no problem type this application has ever heard of.
 *
 * An earlier version listed `upstream-timeout` and `upstream-unusable` and nothing else. Stopping
 * the gateway mid-submission during the live walkthrough produced a bare `500` from the proxy in
 * front of it, and the screen said **"Not sent"** about a transfer that may well have posted -
 * which is precisely the statement WP-14's constraints forbid, and precisely the one that gets a
 * customer to send their money a second time.
 */
export function isOutcomeUnknown(problem: Problem): boolean {
  return problem.status >= 500;
}
